Redesigned learner view by adding Information Card and its metadata

MAKOSCAFEE opened the pull request:
Hi @seanlip and @amitdeutsch 
This is still the work in progress but I wanted to get feedback on the UI implementation of the mock up. 
Thanks!

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2140,3 +2145,90 @@ md-card.preview-conversation-skin-supplemental-card {
+
+.oppia-modal-information-card .modal-dialog {
+ width: 400px;
+ margin:100px auto;
```
general nits, here and below: (i) single space after colon, (ii) alphabetize attrs within each css rule, (iii) prefix css class names with 'oppia-' to distinguish from bootstrap/material-design/other stuff.

MAKOSCAFEE replied:
Done!

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2140,3 +2145,90 @@ md-card.preview-conversation-skin-supplemental-card {
+  background: #888;
+  border-color: black; 
+}
+.information-card-content .card-metrics li .fa{
```
add single space before {

MAKOSCAFEE replied:
Done!.

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +692,63 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+  $scope.showDialog = function () {
+
+    var modalInstance = $modal.open({
+      animation:true,
```
single space after : 

(Perhaps we should find a way to try and fix these automatically.)

MAKOSCAFEE replied:
Done!

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -19,11 +19,13 @@
       <span class="oppia-navbar-breadcrumb-separator"></span>
       <span class="oppia-exploration-title protractor-test-exploration-header" itemprop="description">{{exploration_title}}</span>
     </li>
+    <li ng-click="showDialog()" tooltip="Information Card" tooltip-placement="bottom">
```
showInformationCard()

MAKOSCAFEE replied:
Done!

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +107,83 @@
+          <div flex="55">
+            <md-content layout="row">
+              <div class="oppia-tag-icon"><img src="/third_party/static/material-design-icons-1.0.1/ic_tag.png"></div>
+              <md-chips>
```
Something seems a bit off with the indentation and the number of opening/closing tags for all these md-chip(s).

MAKOSCAFEE replied:
Done!.

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +107,83 @@
+                </a>
+              </li>
+              <li>
+                <a href="" ng-click="showEmbedExplorationModal(explorationId)">
```
@amitdeutsch -- think it's ok to show a modal in a modal here?

MAKOSCAFEE replied:
But this modal is needed for those who need embedded links. Do we need to close the first modal before rendering another modal or not display it at all?. @amitdeutsch

seanlip replied:
I was thinking we could do something like slide open a drawer at the bottom of the card (so the card extends with additional information), but I suggest not implementing this until @amitdeutsch has had a chance to comment. In any case, the current version (modal on modal) is fine for now; we can improve it in a separate pass.

seanlip commented:
Thanks, @MAKOSCAFEE -- overall, I like it! I think you did a good job trying to replicate the mock, though @amitdeutsch may have more comments here.

Btw, note that you'll need to specify the material design icons/images needed in `scripts/install_third_party.py`, otherwise we won't be able to see them (the `third_party` folder isn't tracked by git). There's a bunch of images missing at the moment.

Other thoughts:
- The title is a bit close to the boundary, it could be slightly higher.
- @amitdeutsch Should we truncate long descriptions to max two lines, or impose an overall character limit on that field?
- @amitdeutsch what happens when the 4+ button on the right is clicked? Probably a popover appears on-hover, I expect.
- @amitdeutsch might be cleaner to have the tags at the bottom ("Mathematics, Combinatorics", etc.) be restricted to one line (maybe just two tags) and show the full list in a tooltip when hovered over.

MAKOSCAFEE commented:
@seanlip  and @amitdeutsch 

+ i have added a margin to the title 
+ Next phase is to get description from ```service`` and I think we can truncate to two lines and add ```show more``` which when clicked can expand to full description and ```show less``` to truncate it  again. What do you think?
+ in full functionality 4+ button on-hover should "popover" list of other 4 more profiles
+ I will implement tags to be limited to 2 tags and display others on hover
Thanks for feedback I will work on the comments.

MAKOSCAFEE commented:
Hi @seanlip  can you take a look at this!?. 
I am not sure if I got the APIs right.  I have added the drop down to a button  which shows there are more contributors which and Added tooltips to "4+" chip of tags and coloured it blue.
Thanks!

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,70 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+        $scope.explorationCardTitle = oppiaPlayerService.getExplorationTitle();
+        $scope.lastModifiedDate = _informationCardData.lastTimeUpdated;
+
+        var viewersNumberFilter = function (viewersNumber) {
```
Suggest pulling this filter out to filters.js and adding tests for it in filtersSpec.js

MAKOSCAFEE replied:
Done!

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -362,6 +406,18 @@ oppia.factory('oppiaPlayerService', [
       return _explorationId;
     },
+    getInformationCardData: function() {
+      informationCardData = {
```
Should have a 'var' (or you can just remove the intermediate variable and return the object directly).

MAKOSCAFEE replied:
my bad. I have added ```var```

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,70 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+        };
+        $scope.explorationRatings = ratingService.getUserRating() !== null ? ratingService.getUserRating() : 0;
+        $scope.explorationTags = _informationCardData.explorationTags;
+        $scope.explorationDescription = _informationCardData.explorationObjective;
```
Can we capitalize the first letter? We already do this in the gallery; it might be a good idea to extract the logic into a common filter and put it in filters.js.

MAKOSCAFEE replied:
Sure I have made a common filter which takes in Objective and number of character ```MAX```

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -19,11 +19,13 @@
       <span class="oppia-exploration-title protractor-test-exploration-header" itemprop="description">{{exploration_title}}</span>
     </li>
+    <li ng-click="showInformationCard()" tooltip="Information Card" tooltip-placement="bottom">
+    <img src="/third_party/static/material-design-icons-1.0.1/ic_info_white_48dp.png" class="oppia-navbar-breadcrumb-icon"></li>
```
This should be indented by 2, and the `</li>` moved to the next line. Also, when hovering on this image, the mouse cursor should change to a pointer to indicate that it's clickable.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +107,92 @@
+                </a>
+              </li>
+              <li>
+                <a ng-href="https://plus.google.com/share?url=<[serverName]>/explore/<[explorationId]>" onclick="return !window.open(this.href, '', 'height=600, width=600, menubar=no, toolbar=no, resizable=yes, scrollbars=yes')" target="_window">
```
Surround each of these by the appropriate `if SHARING_OPTIONS[...]` blocks, as is done in `exploration_player.html`. We can also remove the sharing button and the existing code from the navbar, since it duplicates this.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -302,6 +307,43 @@ oppia.factory('oppiaPlayerService', [
+
+      var infoCardDataUrl = '/createhandler/data/' +  _explorationId;
+      var infoCardSnapshotsUrl = '/createhandler/snapshots/' +  _explorationId;
+      var infocardStatisticUrl = '/createhandler/statistics/' +  _explorationId + '/all';
```
infoCardStatisticsUrl

MAKOSCAFEE replied:
Done

seanlip commented:
Hi @MAKOSCAFEE , nice work!

I need to give this another, more careful pass (when I'm a bit more awake), but I've made some comments. I'm also assigning this to @amitdeutsch to take a look at UI.

amitdeutsch commented:
Fantastic job on this, @MAKOSCAFEE !

I just submitted a pull request making numerous font and margin changes to your branch. Feel free to keep or reject any of them.

There are a few other issues that I didn't fix in my pull request, which I'll list here:

![screen shot 2015-10-02 at 9 19 55 am](https://cloud.githubusercontent.com/assets/1188049/10251606/cb4e8cf0-68e6-11e5-9a65-18831e6fcac9.png)

-When hovering over the icons of the contributors, it would be nice to see their user names as a tooltip
-The additional contributors beyond 2 should be listed as '+1', not '1+'
-When I hover over the '+1', I would expect to see the names of other contributors in a tooltip, instead of a floating dropdown with their user icons.
-If there are no contributors (i.e. the exploration is public), I would want that to be communicated, instead of just leaving that area blank. One option for this is to use a 'world' icon to indicate that the exploration is public.

Those are the main things. Overall, fantastic job!

MAKOSCAFEE commented:
Hi @amitdeutsch!
Thanks for feedbacks. I have fixed all issues you mentioned above. 
Thanks!.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -415,20 +415,25 @@ textarea {
 }
 
+.oppia-navbar-breadcrumb-icon {
+  cursor: pointer;
```
Can the cursor: pointer go on the surrounding 'li', so that the cursor changes exactly at the same time when the tooltip appears?

MAKOSCAFEE replied:
Done!

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2144,3 +2150,126 @@ md-card.preview-conversation-skin-supplemental-card {
+  width: 24px;
+}
+
+.oppia-information-card-content .card-metrics li .fa {
```
Put this just below the '.card metrics li" rule, above

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2144,3 +2150,126 @@ md-card.preview-conversation-skin-supplemental-card {
+  margin-right: 5px;
+}
+
+.oppia-social-media-and-tags {
```
Suggest .oppia-info-card-bottom-row.

Reasoning: the 'info-card' gives some indication (like breadcrumbs) about where in the app these rules apply to -- especially since this CSS rule is likely to be specific to just the info card here. If you like, you can similarly change 'oppia-information-card' in the rules above to 'oppia-info-card'.

MAKOSCAFEE replied:
Done. And I  have changed all oppia-information-card to oppia-info-card.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2144,3 +2150,126 @@ md-card.preview-conversation-skin-supplemental-card {
+
+/* This is needed to overide the maximum width of tooltip
+   otherwise  breaks when you have longer text in tooltip */
+.tooltip-inner {
```
This is too broad -- it applies to every .tooltip-inner class in the whole application. Probably need to add additional parent classes.

MAKOSCAFEE replied:
The problem with this is, The tooltips in the oppia-info-card are appended to the body(```tooltip-append-to-body=true```) not relative to the div in the card. This is because tooltip position is relative by default and our divs are small to contain all the content if we have many tags for example try remove this rule and see. And you cant override and set it to absolute or something else so bootstrap provided the way round by appending it to body and stay relative to body.

Otherwise this rule has no side effect to tooltips, as for small tooltips they will still be wrapped around depending on their width.

seanlip replied:
So, I did remove this rule, and nothing seemed to change -- in fact I searched the DOM and couldn't find anything with a `tooltip-inner` class. Is there something specific I should be looking for?

MAKOSCAFEE replied:
add tags from eight and more, while you have removed that rule.. and hover through ```6+ more``` content will be cut.

seanlip replied:
I looked into this. Here's what it looked like for me *without* the tooltip-inner class:

![without-tag](https://cloud.githubusercontent.com/assets/10575562/10329472/dacccfdc-6c76-11e5-9fbd-f17d1d5fd012.png)

I think this is fine, actually, and thus I think it's also fine to remove the .tooltip-inner.

---

On a different note, could you please add a single space immediately after every comma that separates two tags? It's more readable that way.

In addition, try entering a small tag followed by a long one. The tooltip section starts to look a bit weird. Perhaps the truncation logic needs to be adjusted (or alternatively we can also say, it's probably OK for now, since long tags like this are unlikely anyway):

![toomanytags](https://cloud.githubusercontent.com/assets/10575562/10329481/0105a2d2-6c77-11e5-8e3e-e4363ddf8723.png)

MAKOSCAFEE replied:
This is because we don't truncate but we just limit displayed tags to maximum of two tags. So should we truncate if tags are long like that one?. 
That and I am going to remove the inner class .. It looks like appending to the body was just enough. Thanks

seanlip replied:
Hm, interesting. How about: compute lengths of tags (in terms of number of characters) -- and stop if adding a new tag causes the length to pass a certain threshold? Then we would always display complete tags (but it doesn't really matter if it's 1, 2, 3 or more, as long as they fit in the space).

MAKOSCAFEE replied:
So What i have done is to show tags to the maximum of two line meaning(45 characters). I have written a function which checks if tags to be shown exceed that number before is  shown and Those that can not be shown are displayed in the tooltip.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2144,3 +2150,126 @@ md-card.preview-conversation-skin-supplemental-card {
+  width: 28px;
+}
+
+.oppia-social-media-and-tags ul li {
```
Move these near the other .oppia-social-media-and-tags rules

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2144,3 +2150,126 @@ md-card.preview-conversation-skin-supplemental-card {
+  padding: 0;
+}
+
+.oppia-public-owned-icon {
```
.oppia-info-card-community-editable-icon

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2144,3 +2150,126 @@ md-card.preview-conversation-skin-supplemental-card {
+  width: 20px;
+}
+
+.oppia-exploration-contributors-profile {
```
.oppia-info-card-exploration-contributors-profile

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -336,4 +336,34 @@ oppia.filter('summarizeDefaultOutcome', ['$filter', function($filter) {
+  };
+}]);
+
+oppia.filter('formattedObjective', [function() {
```
I think this should be renamed to truncateAndCapitalize, so that it's more generic. At this point the filter doesn't really care that what's being passed to it is an objective.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -336,4 +336,34 @@ oppia.filter('summarizeDefaultOutcome', ['$filter', function($filter) {
+oppia.filter('formattedObjective', [function() {
+  return function(objective, maxNumberOfCharacters) {
+    objective = objective.trim();
+    if(objective.length > maxNumberOfCharacters){
```
nit: add space before {

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -336,4 +336,34 @@ oppia.filter('summarizeDefaultOutcome', ['$filter', function($filter) {
-}])
+}]);
+
+oppia.filter('appendMetrixPrefix', [function() {
```
Rename: 'summarizeNumber'

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -336,4 +336,34 @@ oppia.filter('summarizeDefaultOutcome', ['$filter', function($filter) {
+
+oppia.filter('appendMetrixPrefix', [function() {
+  return function (input) {
+    return Math.abs(Number(input)) >= 1.0e+9
```
We refer to Math.abs(Number(input)) here many times; probably better to define it as a separate variable.

(Though would we ever care about negative numbers? I feel like we should rename this to 'summarizeNonnegativeNumber' and log an error if the number is negative.)

MAKOSCAFEE replied:
sure! changed that. I think number should be checked if is negative or not  before you apply the filter, What do you think? and  for this case of viewers count do we need to log any error since the numbers are positive(>0)?

seanlip replied:
I think that's fine as well -- just add a comment saying

    // Users of this filter should ensure that the input is a non-negative number.

Then you don't need to do any error-checking here.

MAKOSCAFEE replied:
Done!

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -336,4 +336,34 @@ oppia.filter('summarizeDefaultOutcome', ['$filter', function($filter) {
+    return Math.abs(Number(input)) >= 1.0e+9
+
+    // Nine Zeroes for Billions
+    ? (Math.abs(Number(input)) / 1.0e+9).toFixed(1) + "B"
```
nit, here and below: we use single quotes in JS

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+      animation: true,
+      templateUrl: 'popover/informationCard',
+      windowClass: 'oppia-modal-information-card',
+      controller: ['$scope', '$http', '$modal', '$modalInstance','oppiaPlayerService', 'ratingService',
```
nit: add space after comma (after $modalInstance)

Can you make these lines < 100 chars?

MAKOSCAFEE replied:
Done!

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+        $scope.showEmbedExplorationModal = embedExplorationButtonService.showModal;
+
+        $scope.explorationId = oppiaPlayerService.getExplorationId();
+        $scope.infoCardBackgroundImage = {
```
infoCardBackgroundImageCss

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+
+        $scope.explorationId = oppiaPlayerService.getExplorationId();
+        $scope.infoCardBackgroundImage = {
+          'background-image': 'url('+ _informationCardData.infoCardImageUrl +')'
```
nit: add space after '+'

(Are you able to install e.g. jslint/jshint?)

MAKOSCAFEE replied:
Done.
I have installed Jslint now.

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -335,6 +377,7 @@ oppia.factory('oppiaPlayerService', [
             (GLOBALS.CATEGORIES_TO_COLORS[_exploration.category] || 'teal') +
             '_large.png');
           _loadInitialState(successCallback);
+          _loadInitialInformationCardData(successCallback);
```
Is this actually needed in edit mode? If not I'd remove it.

MAKOSCAFEE replied:
I have removed it. 
But  in edit mode the card will be shown without metadatas right?

seanlip replied:
I don't think the card is shown at all in edit mode? The navbar is different.

MAKOSCAFEE replied:
I meant in preview mode!

MAKOSCAFEE replied:
i meant in preview mode

seanlip replied:
Yep, I know. The navbar is still different ... try it!

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -344,6 +387,7 @@ oppia.factory('oppiaPlayerService', [
           sessionId = data.session_id;
           _viewerHasEditingRights = data.can_edit;
           _loadInitialState(successCallback);
+          _loadInitialInformationCardData(successCallback);
```
You'll end up calling successCallback twice if you do this. I don't think you need to pass in successCallback...

MAKOSCAFEE replied:
removed  ```successCallback```

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -302,6 +307,43 @@ oppia.factory('oppiaPlayerService', [
+      var infoCardSnapshotsUrl = '/createhandler/snapshots/' +  _explorationId;
+      var infocardStatisticsUrl = '/createhandler/statistics/' +  _explorationId + '/all';
+
+      // This is needed to get exploration objective/goal,contributors. 
```
nit: add space after comma; remove trailing full stop.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -302,6 +307,43 @@ oppia.factory('oppiaPlayerService', [
 
+  var _loadInitialInformationCardData = function(successCallback) {
+
+      var infoCardDataUrl = '/createhandler/data/' +  _explorationId;
```
Nit: extra space before _explorationId.

Also, add:

    // TODO(sll): move these three handlers into a separate exp_metadata.py controller.

(since now they're being used by both the player and the editor).

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -302,6 +307,43 @@ oppia.factory('oppiaPlayerService', [
+      $http.get(infoCardDataUrl).success(function(data) {
+        _explorationObjective = data.objective;
+        _explorationTags = data.tags;
+        _explorationContributors = data.rights.owner_names.concat(data.rights.editor_names);
```
Add TODO:

  // TODO(sll): compute the actual contributors based on the list of commits in the backend

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -302,6 +307,43 @@ oppia.factory('oppiaPlayerService', [
+        _explorationContributors = data.rights.owner_names.concat(data.rights.editor_names);
+      }).error(function () {
+        warningsData.addWarning(
+            data.error || 'There was an error loading the exploration information.');
```
nit: 2-space not 4-space indentation

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -147,6 +147,11 @@ oppia.factory('oppiaPlayerService', [
+  var _explorationObjective = null;
+  var _explorationTags = [];
+  var _explorationContributors = [];
+  var _lastTimeUpdated = null;
```
Better to call this something like

    _explorationLastUpdatedMsec

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -302,6 +307,43 @@ oppia.factory('oppiaPlayerService', [
+      // This is needed to get exploration snapshots.
+      $http.get(infoCardSnapshotsUrl).success(function(response) {
+        var data = response.snapshots;
+        //only get last published chages/get changes from the last snapshot.
```
nits:
- space after "//"
- s/chages/changes

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -302,6 +307,43 @@ oppia.factory('oppiaPlayerService', [
+
+      var infoCardDataUrl = '/createhandler/data/' +  _explorationId;
+      var infoCardSnapshotsUrl = '/createhandler/snapshots/' +  _explorationId;
+      var infocardStatisticsUrl = '/createhandler/statistics/' +  _explorationId + '/all';
```
infoCard...

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -147,6 +147,11 @@ oppia.factory('oppiaPlayerService', [
   var _infoCardImageUrl = null;
+  var _explorationObjective = null;
+  var _explorationTags = [];
+  var _explorationContributors = [];
```
_explorationContributorUsernames (or _explorationAuthorUsernames)

MAKOSCAFEE replied:
I choose to use  ```_explorationContributorUsernames```
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+        $scope.explorationDescription = _informationCardData.explorationObjective;
+        $scope.explorationContributors = _informationCardData.explorationContributors;
+        $scope.explorationCardTitle = oppiaPlayerService.getExplorationTitle();
+        $scope.lastModifiedDate = _informationCardData.lastTimeUpdated;
```
Try and reuse whatever name has already been used in _informationCardData, to prevent having multiple names for the same 'thing'.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+        $scope.explorationCardTitle = oppiaPlayerService.getExplorationTitle();
+        $scope.lastModifiedDate = _informationCardData.lastTimeUpdated;
+
+        $scope.viewsCount = _informationCardData.viewersCount;
```
Ditto.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+        };
+        $scope.cancel = function () {
+          $modalInstance.dismiss();
+        }
```
Missing semicolon.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+        $scope.onMouseoutDropdownMenu = function(evt) {
+          angular.element(evt.currentTarget).parent().removeClass('open');
+        };
+        $scope.cancel = function () {
```
Here and below: no space after 'function'

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +65,100 @@
+          <li>
+            <span class="fa fa-eye fa-lg"></span><[viewsCount | summarizeNumber]></li>
+          <li>
+            <span class="fa fa-clock-o fa-lg"></span><[lastModifiedDate | date: 'MM/dd/yyyy']></li>
```
You might want to use oppiaDatetimeFormatter instead (see app.js).

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +65,100 @@
+            <img src="/third_party/static/material-design-icons-1.0.1/ic_group_black_48dp.png">
+            <li ng-repeat="contributor in explorationContributors| limitTo: 2"
+              tooltip="<[contributor]>" tooltip-placement="top">
+              <img class="img-circle" ng-src="/images/general/user_mint_48px.png" alt=""></li>
```
Bring `</li>` down to next line. (Only keep it on the same line if the `<li>` is on that line too, otherwise the indentation looks weird.)

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +65,100 @@
+            <li ng-if='explorationContributors.length > 2' class="oppia-contributors-more-circle"
+              tooltip-append-to-body=true tooltip="<[explorationContributors.slice(2).join()]>"
+              tooltip-placement="top">+<[explorationContributors.length - 2]></li>
+            <li ng-if='explorationContributors.length < 1' tooltip="public" tooltip-placement="top">
```
=== 0?

MAKOSCAFEE replied:
Done!

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +65,100 @@
+          </div>
+          <div flex="45">
+            <ul layout="row" layout-align="end center">
+              <li>
```
I spoke to Amit -- let's get rid of this `<li>` entirely, since it duplicates the embed link at the end. The G+ etc. icons should be obvious enough.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+        $scope.lastModifiedDate = _informationCardData.lastTimeUpdated;
+
+        $scope.viewsCount = _informationCardData.viewersCount;
+        $scope.onMouseoverDropdownMenu = function(evt) {
```
Are these functions being used (onMouseover and onMouseout)?

MAKOSCAFEE replied:
Done

seanlip commented:
@MAKOSCAFEE -- this is really good, well done! I have a bunch of code-related comments but once these are sorted out I think this can go in.

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +748,53 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+          GLOBALS.SHARING_OPTIONS_TWITTER_TEXT);
+        $scope.showEmbedExplorationModal = embedExplorationButtonService.showModal;
+
+        $scope.explorationId = oppiaPlayerService.getExplorationId();
```
I think we need to add a TODO to turn both getExplorationId() and getExplorationTitle() into promises. The reason is, I loaded the page and immediately clicked on the info icon. This gave a console error since the title wasn't fetched yet. We probably should refactor so that these methods return promises instead.

It's fine to do this in a separate pass, though -- it'll require changes in other places and I don't think we should make the scope of this PR any larger. Hence the suggestion to add a TODO.

MAKOSCAFEE replied:
Done

seanlip commented:
Btw, @MAKOSCAFEE, I also meant to say that I think you did an extremely good job capturing the spirit of the mocks, and that the info card looks really nice. Well done, and thanks a lot!

MAKOSCAFEE commented:
Hi @seanlip  Thanks.
I want to inform you that all review comments have been addressed.
Thanks.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2145,6 +2150,129 @@ md-card.preview-conversation-skin-supplemental-card {
+  font-family: "Capriola", "Roboto", Arial, sans-serif;
+  margin: 12px 0;
+  padding: 0;
+  position:absolute;
```
nit: add space after colon

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -336,4 +336,41 @@ oppia.filter('summarizeDefaultOutcome', ['$filter', function($filter) {
-}])
+}]);
+
+// Filter that summarizes large number to nearly four significant figures and
```
Filter that summarizes a large number to a decimal followed by
the appropriate metric prefix (K, M or B). For example, 167656
becomes 167.7K.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/filtersSpec.js:
```diff
@@ -180,4 +182,37 @@ describe('Testing filters', function() {
+
+    expect(filter('remove New line', 4)).toEqual('Remo...');
+    // If maximum number of character are not specified
+    // return whole objective but first letter capitalize.
```
// If the maximum number of characters is not specified, return
// the whole objective with the first letter capitalized.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/filtersSpec.js:
```diff
@@ -180,4 +182,37 @@ describe('Testing filters', function() {
+    expect(filter('remove New line', 4)).toEqual('Remo...');
+    // If maximum number of character are not specified
+    // return whole objective but first letter capitalize.
+    expect(filter('capitalize first letter and trancate')).toEqual('Capitalize first letter and trancate');
```
s/trancate/truncate

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/filtersSpec.js:
```diff
@@ -180,4 +182,37 @@ describe('Testing filters', function() {
+
+  }));
+
+  it('should capitalize first letter and truncate string ', inject(function($filter) {
```
nit: no space before closing quote

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -302,6 +307,46 @@ oppia.factory('oppiaPlayerService', [
+      _explorationTags = data.tags;
+      // TODO(sll): in the backend, compute the actual contributors
+      // based on the list of commits to the exploration.
+      _explorationContributorUsernames = data.rights.owner_names.concat(data.rights.editor_names);
```
In JS please try and keep line lengths to 80 chars where possible (failing that, 100 chars). This should be possible to set in the JSLint settings.

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -692,3 +750,54 @@ oppia.directive('feedbackPopup', ['oppiaPlayerService', function(oppiaPlayerServ
+          'background-image': 'url('+ _informationCardData.infoCardImageUrl + ')'
+        };
+        $scope.explorationRatings = ratingService.getUserRating() !== null ? 
+                                    ratingService.getUserRating() : 0;
```
Should probably be 'Unrated' (maybe italicized) if there are no ratings yet

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +65,106 @@
+        <p ng-if="!explorationDescription">No objective specified.</p>
+        <ul layout="row" class="card-metrics" layout-align="space-between center">
+          <li>
+            <span class="fa fa-star fa-lg"></span>
```
Can you give these icons tooltips? Ratings, Views, Last Updated, Contributors

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +65,106 @@
+            <md-content layout="row">
+              <div class="oppia-info-card-tag-icon"><span class="fa fa-tags fa-lg"></span></div>
+              <md-chips layout-align="center center">
+                  <md-chip ng-if='explorationTags.length > 2' class="oppia-info-card-tooltip-more"
```
indentation

MAKOSCAFEE replied:
Done

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +65,106 @@
+                  <md-chip ng-if='explorationTags.length > 2' class="oppia-info-card-tooltip-more"
+                    tooltip-append-to-body=true tooltip="<[explorationTags.slice(2).join()]>"
+                    tooltip-placement="right">
+                    <[explorationTags.length - 2]>+ more tags
```
I don't understand why hovering over this is causing the cursor to change to a 'text input' cursor. It should probably just stay the same...

MAKOSCAFEE replied:
I had forgotten, chips are primary inputs that you can add tags dynamically. I have overridden that property of the cursor

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -105,6 +65,106 @@
+            <img src="/third_party/static/material-design-icons-1.0.1/ic_group_black_48dp.png">
+            <li ng-repeat="ContributorUsername in explorationContributorUsernames| limitTo: 2"
+              tooltip="<[ContributorUsername]>" tooltip-placement="top">
+              <img class="img-circle" ng-src="/images/general/user_mint_48px.png" alt="">
```
Can these be a little smaller (say 20px by 20px) while keeping the ic_group_black_48dp.png icon at the same size as the others (star, eye, etc.)? The user icons should feel subsidiary to the 'header' icon.

MAKOSCAFEE replied:
Done.

seanlip commented:
Hi @MAKOSCAFEE, I've done a fairly careful pass. Generally, I think it looks very nice! Once the above comments are addressed we should be able to merge it.

MAKOSCAFEE commented:
Hi @seanlip  I addressed all review comments. Sorry I have had alot of nit and indentation errors. I installed JSLint and I am trying to customize it to go with standard of this codebase.
Thanks!

seanlip commented:
Thanks @MAKOSCAFEE! 

Just took another look at UI. Some notes:

- Let's get rid of the italic tags around "Unrated" -- it looks a bit off. Sorry for the bad suggestion.
- Change the tooltip for the globe icon to "Community editable" (no hyphen)
- When there are no tags, show `<em>No tags specified.</em>`
- Add a 'Tags' tooltip for consistency.
- When the data hasn't loaded from the backend yet (e.g. load the page and immediately click the 'i' icon), show a loading message on the bottom part of the card until the data has loaded. Otherwise you'll see weird things like the last update time being 31/12/1969.

Sorry these are separate notes -- I only looked at the code just now but didn't get a chance to play with the UI. These should be the last of my comments, though!

MAKOSCAFEE commented:
Hi @seanlip. I have addressed all the additional notes above. Though keeping track of all three ```get``` function was tricky so I used chaining promise and load data when all data are a loaded but this is little bit slower  since functions are executed one after another. Any suggestion for better way to handle this?!
Thanks.

seanlip commented:
Yep, sequential execution here isn't great, as you point out. How about doing them together, so the time needed is the maximum of the per-promise times (instead of the sum)? I believe $q.all can be used for this:

    https://docs.angularjs.org/api/ng/service/$q
    https://www.jonathanfielding.com/combining-promises-angular/

Make sure you handle the error case correctly (e.g. show "Oops, something went wrong. Could not load data." on the card, and call $log.error()).

MAKOSCAFEE commented:
Thanks @seanlip .
Now is faster than before.

seanlip commented:
Looks good! Couple small things, but they're trivial. Thanks, @MAKOSCAFEE!

I suggest fixing these, and then once the Travis tests pass, I'll merge.

MAKOSCAFEE commented:
@seanlip  Addressed review comments. They are fixed.
Regards.
